normits_demand/models/forecasting/EDGE_growth/run.py

```diff
@@ -0,0 +1,342 @@
+# -*- coding: utf-8 -*-
+"""
+Created on: 17/05/2023
+Updated on:
+
+Original author: Ben Taylor
+Last update made by:
+Other updates made by:
+
+File purpose:
+
+"""
+# Built-Ins
+import pathlib
+import sys
+import pickle
+
+sys.path.append(r"E:\NorMITs-Demand")
+# Third Party
+import logging
+import pandas as pd
+import numpy as np
+from tqdm import tqdm
+from caf.toolkit import concurrency
+from functools import partial
+
+# Local Imports
+# pylint: disable=import-error,wrong-import-position
+# Local imports here
+import growth_matrices
+import apply_growth
+import utils
+import loading
+from normits_demand.utils import timing
+from normits_demand.matrices import omx_file
+from normits_demand.utils import file_ops
+from normits_demand.models.forecasting import forecast_cnfg
+
+# pylint: enable=import-error,wrong-import-position
+
+# # # CONSTANTS # # #
+logging.basicConfig(level=logging.INFO)
+LOG = logging.getLogger(__name__)
+# # # CLASSES # # #
+
+# # # FUNCTIONS # # #
+def _tp_loop(
+    factored_matrices,
+    matrices_to_grow_dir,
+    model_stations_tlcs,
+    forecast_year,
+    demand_segments,
+    filled_growth_matrices,
+    growth_summary,
+    time_period,
+    splitting_matrices,
+):
+    """
+    Private function only called once for multiprocessing. All args are as named when read in.
+    """
+    overall_not_grown_demand = 0
+    overall_base_demand = 0
+    LOG.info(
+        "-- Processing Time Period %s @ %s",
+        time_period,
+        timing.get_datetime(),
+    )
+    # read time period specific files
+    irsj_props = pd.read_hdf(
+        matrices_to_grow_dir / f"{time_period}_iRSj_probabilities.h5",
+        key="iRSj",
+    )
+    LOG.info(
+        f"{'Time_Period':>12}{'Demand_Segment':>15}"
+        f"{'Base_Demand':>12}{f'{forecast_year}_Demand':>12}"
+    )
+
+    with omx_file.OMXFile(
+        pathlib.Path(
+            matrices_to_grow_dir,
+            f"PT_{time_period}.omx",
+        )
+    ) as omx_mat:
+        for row in tqdm(
+            demand_segments.itertuples(),
+            desc="    Demand Segments Loop ",
+            unit=" Segment",
+            colour="cyan",
+            total=len(demand_segments),
+        ):
+            # store current segment's details
+            segment = row.Segment
+            to_home = row.ToHome
+            growth_method = row.Growth_Method
+            userclass = row.Userclass
+            purpose = row.Purpose
+            # read demand matrix
+            zonal_base_demand_mx = utils.wide_mx_2_long_mx(
+                omx_mat.get_matrix_level(segment),
+                rows="from_model_zone_id",
+                cols="to_model_zone_id",
+            )
```

Comment:
There's some functionality in caf.toolkit to do this - it would be worth digging out everywhere this is used and replacing it with that functionality. Lets not repeat code if we don't have to.

Reply:
Looks like I've been going through comments in the wrong order! Addressed elsewhere and I hopefully will have changed all of these by the time you read this!